### DISCUSSION TITLE
How are channel entries updated in jamulus.ini ?

### DISCUSSION THREAD

**Discussion by pcar75** [jamulussoftware/jamulus#3063] [Technical Support & Help] [answered]
Bonjour,

I suggested some time ago ("Remember specially marked faders #2842")  a modification to prioritise preservation of fader entries that are set to muted in jamulus.ini, as with time, entries are purged to make place for new ones , although I do not undestand specifically how.

For my part, I sought to export from jamulus.ini  muted fader entries to consolidate them in a single file ( *.jch) in the hope of occasionaly reintegrate them in the jamulus.ini via the "Load Mixer Channels setup ..." (once the Jamulus client exits and rewrites the jamulus.ini with then current values).

After some prototyping (AWK programs) and tests , I noticed that the "Load Mixer Channels Setup ..." corrupts the jamulus.ini file upon exit.

Test in Windows10 with Jamulus v3.9.1
Procedure : start Jamulus client, stay unconnected, « Load Mixer Channels Setup... » with Jamulus_Muted.jch , exit Jamulus Client.

    Jamulus_Muted.jch contains only 10 full entries, muted, tagged from 1 to 10 (not the original tag numbers).

    JamulusW.ini Before test contains 250 entries of which 3 are muted , one of these already present under another tag number.

    JamulusW.ini After test contains 4 muted entries, some lost, some not muted, some duplicated.

My C is Ok, but I have a lot ground to cover before familiarity with C++ and Qt libraries/Framework.

1.	Could a developper tell me how exactly channel entries in jamulus.ini are updated from session to session and from the use of « Load Mixer Channels Setup ... » ?

3.	Should this be a bug fix request ( to inhibit duplication Alias entries, favoring update from or creation of new entries ) ?

Thanks in advance.

**Comment by pljones** [comment]
1. The serialization support is in settings.cpp:
https://github.com/jamulussoftware/jamulus/blob/main/src/settings.cpp#L216-L226
https://github.com/jamulussoftware/jamulus/blob/main/src/settings.cpp#L810-L834
`vecStoredFaderTags` is the interesting element, really.  This is managed here
https://github.com/jamulussoftware/jamulus/blob/main/src/audiomixerboard.cpp#L1672 with the most important lines being
https://github.com/jamulussoftware/jamulus/blob/main/src/audiomixerboard.cpp#L1684-L1691
2. I'd have expected
https://github.com/jamulussoftware/jamulus/blob/main/src/audiomixerboard.cpp#L1684
to have the effect of ensuring there was only one entry for a given name.

**Reply by pcar75** [reply]
@pljones Thanks.  I will try to understand these parts of code.
I think the problem is with the "Load Mixer Channels Setup".
I've attached my AWK scripts (renamed as .txt) so you can try on your own.
I could send my Jamulus.ini.bak (before), JamulusMuted.jch, Jamulus.ini (after) and my result report but it's not appropriate to post my "muteds"  publicly in this forum.
If these are deemed necessary , where and how do I send them ?
Thanks in advance.

[JListMutedawk.txt](https://github.com/jamulussoftware/jamulus/files/11466974/JListMutedawk.txt)
[JExportMutedawk.txt](https://github.com/jamulussoftware/jamulus/files/11466976/JExportMutedawk.txt)
[JListAllawk.txt](https://github.com/jamulussoftware/jamulus/files/11525076/JListAllawk.txt)

EDIT : modified ListAllawk.txt to add muted flag ('M') and correct muted count.

**Reply by ann0see** [reply]
> but it's not appropriate to post my "muteds" publicly in this forum.
If these are deemed necessary , where and how do I send them ?

There's no PM feature in GitHub, but you could use E-Mail - if Peter is ok with it

**Reply by pljones** [reply]
I'll not have time to look at anything immediately - that is, I've no idea when I will have time.

**Comment by pcar75** [comment]
@pljones  I'm looking at a version I had downloaded a while ago ( jamulus-r3_9_1rc1 ) and the line numbers you indicated ( audiomixerboard.cpp ) seem off a bit ?!?  Is this the version you were referring to ?

**Reply by pljones** [reply]
No.  I'm referring to `HEAD` of `main` at the time I wrote my message.  Things could have changed since...

**Reply by pcar75** [reply]
Bonjour, I do not understand what your reply means.
I see no file named HEAD and in main.cpp there are no token named HEAD, only precompiler directives with HEADLESS.
Could you clarify, please ?
  
Just in case, I've redownloaded a source version which unless mistaken is the actual stable 3.9.1.
https://github.com/jamulussoftware/jamulus/releases/tag/r3_9_1 > Assets > Source code (zip) Oct 17, 2022
and will base further probing on this std version.

My understanding is 
settings.cpp#L216-L226 refers to function CClientSettings::SaveFaderSettings()
settings.cpp#L810-L834 refers to function CClientSettings::WriteFaderSettingsToXML ()

audiomixerboard.cpp#L1672 refers to statement CVector<int> vbOldStoredFaderIsSolo ( pSettings->vecStoredFaderIsSolo );
audiomixerboard.cpp#L1684 refers to statement  pSettings->vecStoredFaderGroupID[0] = pChanFader->GetGroupID();
    but I guess the overall intention was to refer to function 
audiomixerboard.cpp#L1665-L1690 (not L1684-LL691) CAudioMixerBoard::StoreFaderSettings ( )

Thanks for your guidance.

**Reply by pljones** [reply]
https://git-scm.com/doc is a good place to start for git terminology.

**Reply by ann0see** [reply]
Yes. HEAD is a git terminology. That's basically the latest state of the code (available and accepted by the team - however currently it's not yet in a stable release)

I'd suggest you to install and get used to git. For a quick and dirty quick start you can read the CONTRIBUTING.md file and read something like https://medium.com/nebula-graph/how-to-be-a-github-contributor-in-five-minutes-72175c036c17
The linked book is also a good reference.

**Comment by pcar75** [comment]
I will be doing some other tests, but that function "Load Mixer Channels Setup..." is puzzling.  Here are the detailed results of the test I referred to in the first place (best viewed with not-proportionnal font, e.g. courrier new).
[TestResult_anonymized.txt](https://github.com/jamulussoftware/jamulus/files/11512796/TestResult_anonymized.txt)

**Reply by dcorson-ticino-com** [reply]
A quick look at the code and I see that the .jch read and write routines are very simplistic and do no error checking.
They assume that all 256 possible entries are written and read every time.
If you are creating a .jch file and expecting this to be integrated in the current list it will not work.
To set mute values automajically your best best may be to do that offine in your ini file.
Search for the base64 tag of the person you want to manipulate and change the mute value of that entry.

**Reply by pcar75** [reply, marked as answer]
@dcorson-ticino-com Thanks ,  I was arriving at that conclusion too, from some tests in "black-box" mode.
But since those "import" and "export" function exists, I feel I would be sad not to exploit their capabilities better,
I will make this as a feature request then.

[EDIT] Ref.: Issues 
 Enhance Load Mixer Channels #3071 
 Enhance Import/Export. #2831 

Note ...
> They assume that all 256 possible entries

There is a maximum of 250 entries, including owner  ;-)

**Comment by pcar75** [comment]
If it helps anybody, here are AWK prototypes of what I had in mind for preserving muted entries.

- Collect muted entries from jamulus.ini and create/update JAllMuted.jch
[JCollectMuted.awk.txt](https://github.com/jamulussoftware/jamulus/files/11691865/JCollectMuted.awk.txt)

- Import muted entries collected in JAllMuted.jch and from jamulus.ini  into JamulusTemp.ini 
[JImportMuted.awk.txt](https://github.com/jamulussoftware/jamulus/files/11691872/JImportMuted.awk.txt)

I went for clarity , not AWKish performance or proficiency. Nonetheless . they could provide guidance  if / when better use of the import / export functions of Jamulus is developped .

**Reply by ann0see** [reply]
Feel free to add them to the community knowledge base - or even better - upload them to a repository. https://jamulus.io/kb/index.html
